internal/job/executor.go

```diff
@@ -1717,48 +1718,64 @@ func (e *Executor) runPostCommandHooks(ctx context.Context) error {
 }
 
 // CommandPhase determines how to run the build, and then runs it
-func (e *Executor) CommandPhase(ctx context.Context) (error, error) {
+func (e *Executor) CommandPhase(ctx context.Context) (hookErr error, commandErr error) {
+	var preCommandErr error
+
 	span, ctx := tracetools.StartSpanFromContext(ctx, "command", e.ExecutorConfig.TracingBackend)
-	var err error
-	defer func() { span.FinishWithError(err) }()
+	defer func() {
+		span.FinishWithError(hookErr)
+	}()
+
+	// Run postCommandHooks, even if there is an error from the command, but not if there is an
+	// error from the pre-command hooks. Note: any post-command hook error will be returned.
+	defer func() {
+		if preCommandErr != nil {
+			return
+		}
+		hookErr = e.runPostCommandHooks(ctx)
+	}()
+
 	// Run pre-command hooks
-	if err := e.runPreCommandHooks(ctx); err != nil {
-		return err, nil
+	if preCommandErr = e.runPreCommandHooks(ctx); preCommandErr != nil {
+		return preCommandErr, nil
 	}
 
-	// Run the actual command
-	commandExitError := e.runCommand(ctx)
-	var realCommandError error
+	// Run the command
+	commandErr = e.runCommand(ctx)
 
-	// If the command returned an exit that wasn't a `exec.ExitError`
-	// (which is returned when the command is actually run, but fails),
-	// then we'll show it in the log.
-	if shell.IsExitError(commandExitError) {
-		if shell.IsExitSignaled(commandExitError) {
-			e.shell.Errorf("The command was interrupted by a signal")
-		} else {
-			realCommandError = commandExitError
-			e.shell.Errorf("The command exited with status %d", shell.GetExitCode(commandExitError))
-		}
-	} else if commandExitError != nil {
-		e.shell.Errorf(commandExitError.Error())
-	}
+	// Save the command exit status to the env so hooks + plugins can access it. If there is no
+	// error this will be zero. It's used to set the exit code later, so it's important
+	e.shell.Env.Set(
+		"BUILDKITE_COMMAND_EXIT_STATUS",
+		fmt.Sprintf("%d", shell.GetExitCode(commandErr)),
+	)
 
-	// Expand the command header if the command fails for any reason
-	if commandExitError != nil {
-		e.shell.Printf("^^^ +++")
+	// Exit early if there was no error
+	if commandErr == nil {
+		return nil, nil
 	}
 
-	// Save the command exit status to the env so hooks + plugins can access it. If there is no error
-	// this will be zero. It's used to set the exit code later, so it's important
-	e.shell.Env.Set("BUILDKITE_COMMAND_EXIT_STATUS", fmt.Sprintf("%d", shell.GetExitCode(commandExitError)))
+	// Expand the job log header from the command to surface the error
+	e.shell.Printf("^^^ +++")
```

Comment:
This used to happen after then error logs. I don't think it will make a difference to the logs after terminal-to-html is applied. In both cases, the log group of the command is expanded. But obviously, some lines in the raw log will be permuted.

Reply:
i initially thought that there was a change in behaviour here where if pre-command hooks failed, their log groups wouldn't expand, but it seems like that was the case before anyway... kinda odd, but agree that a noop refactor is a good idea.

Reply:
Yeah, there's a few quirks like this. Probably crept in because the old code was hard to reason with.